### PR TITLE
use proj4 string for ST_transform

### LINKED ISSUE #63: Performance Issues in Non-trivial Datasets not in epsg:4326
Hi! First off, excellent project, I enjoyed taking it for a spin! I wanted to let you know that I believe there might be a bit of an opportunity to make the project a bit faster for databases that are rather large and not already in lat lon projection.

When I tried this out recently I pointed it at a fairly large table (350m rows of polygons) which had its geometry already in epsg:3857 (spherical mercator). I noticed that even when requesting z16 tiles out in the ocean where I had no polygons, the request would run pretty much forever (I didn't wait to see how long but it was 10s of minutes). First I checked my indexes and they looked fine. Then I started a request going and checked out `pg_stat_activity` and saw this:

```
 WITH                                       
 bounds AS (                                
     SELECT                                 
         ST_Segmentize(                     
             ST_MakeEnvelope(               
                 $1,                        
                 $2,                        
                 $3,                        
                 $4,                        
                 $5                         
             ),                             
             $6                             
         ) AS geom                          
 ),                                         
 mvtgeom AS (                               
     SELECT ST_AsMVTGeom(                   
         ST_Transform(t.wkb_geometry, $5),  
         bounds.geom,                       
         $7,                                
         $8                                 
     ) AS geom, label, value, ogc_fid       
     FROM public.polygons_14 t, bounds      
     WHERE ST_Intersects(                   
         ST_Transform(t.wkb_geometry, 4326),
         ST_Transform(bounds.geom, 4326)    
     ) LIMIT 10000                          
 )                                          
 SELECT ST_AsMVT(mvtgeom.*) FROM mvtgeom 
```

I noticed that for some reason the `where` clause is always reprojecting the tables geometry to lat lon projection before doing the intersection! I didn't `explain` the query, but instead I rewrote it to stay in spherical mercator the whole time and the querries were blazing fast.

I'm wondering if this was an oversight or if there is maybe an environment variable that i missed setting to tell it not to do this or some other mistake on my part. Just in case it wasnt my fault I thought I should let you know what I ran into.

Thanks again for the wonderful project!

### LINKED ISSUE #65: No tile produce when Tile Matrix Set is WorldCRS84Quad
To query for tiles when `Tile Matrix Set` is set to `WorldCRS84Quad`, there  is no tile produced. It is cause by the `None` value which is returned from `tms.crs.to_epsg()`.  Actually, the `None` value is produced from package `morecantile`, as we can see using the code below.

```python
tms = morecantile.tms.get("WorldCRS84Quad")
print(tms.crs.to_epsg())
# None
```

To avoid this error, I change the code as following diff result for the file `timvt/layer.py`:

```diff
diff --git a/timvt/layer.py b/timvt/layer.py
index 2ab3460..af51b5e 100644
--- a/timvt/layer.py
+++ b/timvt/layer.py
@@ -150,13 +150,15 @@ class Table(Layer):
                 )
                 SELECT ST_AsMVT(mvtgeom.*) FROM mvtgeom
             """
+            epsg = tms.crs.to_epsg()
+            if epsg is None: epsg = 4326
             return await conn.fetchval_b(
                 sql_query,
                 xmin=bbox.left,
                 ymin=bbox.bottom,
                 xmax=bbox.right,
                 ymax=bbox.top,
-                epsg=tms.crs.to_epsg(),
+                epsg=epsg,
                 seg_size=segSize,
                 tile_resolution=int(resolution),
                 tile_buffer=int(buffer),
```

### PR DESCRIPTION
**Pull request by vincentsarago**
closes #65, closes #63 

This PR does:
- compare tile geometry and table geometries in table CRS (#63)
- use proj4 string in ST_Transform when TMS's CRS doesn't have a valid EPSG code (#65)
- update morecantile requirement
- update pre-commit config
- add more tests (for TMS support)

### TODO
- [ ] add a note in the documentation about performances when using non-epsg CRS

### CODE REVIEW

File: timvt/layer.py [hunk 1]
```diff
@@ -129,23 +133,22 @@ async def get_tile(
...
+                        ST_Transform(t.{geometry_column}, {out_proj}),
                         bounds.geom,
                         :tile_resolution,
                         :tile_buffer
                     ) AS geom, {colstring}
                     FROM {self.id} t, bounds
                     WHERE ST_Intersects(
                         ST_Transform(t.{geometry_column}, 4326),
-                        ST_Transform(bounds.geom, 4326)
+                        ST_Transform(bounds.geom, {tms_proj}, 4326)
```
Comment (vincentsarago) [inline timvt/layer.py:151]: Here we use `geometry ST_Transform(geometry geom, text from_proj, integer to_srid);`  (note: we remove the SRID in the `ST_MakeEnvelope` call so we need to add the input projection here)

File: timvt/layer.py [hunk 2]
```diff
@@ -119,9 +120,24 @@ async def get_tile(
...
+
+        # SQL Expression to transform tile bounds in table's geometry CRS
+        if tms_epsg:
+            st_trans_expr = f"ST_Transform(bounds.geom, ST_SRID(t.{geometry_column}))"
+        else:
+            # When there is no EPSG code for the TileMatrixSet we have to use `ST_Transform(geom, in_proj, srid)`
+            st_trans_expr = (
+                f"ST_Transform(bounds.geom, {out_srid}, ST_SRID(t.{geometry_column}))"
+            )
+
```
Comment (vincentsarago) [inline timvt/layer.py:136]: TODO: review

File: timvt/layer.py [hunk 3]
```diff
@@ -130,33 +146,35 @@ async def get_tile(
                                 :ymin,
                                 :xmax,
                                 :ymax,
-                                :epsg
+                                -- If EPSG is null we set it to 0
+                                {tms_epsg or 0}
```
Comment (vincentsarago) [inline timvt/layer.py:150]: TODO: review

File: timvt/layer.py [hunk 4]
```diff
@@ -208,6 +226,12 @@ async def get_tile(
         **kwargs: Any,
     ):
         """Get Tile Data."""
+        # We only support TMS with valid EPSG code
+        if not tms.crs.to_epsg():
+            raise MissingEPSGCode(
+                f"{tms.identifier}'s CRS does not have a valid EPSG code."
+            )
```
Comment (vincentsarago) [inline timvt/layer.py:233]: because we state that function have to accept `epsg` as input, we must raise an exccption if there is no epsg code!

**Comment by vincentsarago** [comment]
latest commit fix most of the performance BUT it made it worse for non-epsg based CRS 

here is what really needs to be reviewed

https://github.com/developmentseed/timvt/blob/a0794ac1b498658c9856d42433e6e17e0e1fcd79/timvt/layer.py#L128-L181

**Comment by vincentsarago** [comment]
To be honest I can't think of a better solution to support `non-epsg` CRS. I think I'll add a note about the performance implication of using such CRS.

**Review by bitner — APPROVED** [review]

**Pull request merged**